tools/lsp/main.rs

```diff
@@ -302,8 +313,19 @@ fn main_loop(connection: Connection, init_param: InitializeParams, cli_args: Cli
         Poll::Pending => futures.push(first_future),
     };
 
-    for msg in &connection.receiver {
-        match msg {
+    let connections = [connection.receiver.clone(), injector_receiver.clone()];
+    let mut selector = {
+        let mut tmp = crossbeam_channel::Select::new();
+        for c in &connections {
+            tmp.recv(c);
+        }
+        tmp
+    };
```

Comment:
This seems a bit convoluted code.

Why not use the `crossbeam_channel::select!` macro.

```rust
loop {
   let msg = select! [
      recv(connection.receiver) -> msg => {
         let Some(msg) = msg else { return Ok(()) }
         msg
      } 
      recv(injector_receiver) -> msg => msg?,
   ] {
```

In particular, we should return Ok when the stdin channel gets disconnected